logstash_async/formatter.py

```diff
@@ -199,17 +229,68 @@ def _move_extra_record_fields_to_prefix(self, message):
         if not self._extra_prefix:
             return  # early out if no prefix is configured
 
-        field_skip_list = constants.FORMATTER_LOGSTASH_MESSAGE_FIELD_LIST + [self._extra_prefix]
+        message.setdefault(self._extra_prefix, {})
+        field_skip_list = self.formatter_logstash_message_field_list + [self._extra_prefix]
         for key in list(message):
             if key not in field_skip_list:
                 message[self._extra_prefix][key] = message.pop(key)
 
+    # ----------------------------------------------------------------------
+    def _post_process_message(self, message):
+        """Override when needed"""
+
     # ----------------------------------------------------------------------
     def _serialize(self, message):
         return json.dumps(message, ensure_ascii=self._ensure_ascii)
 
 
+class LogstashEcsFormatter(LogstashFormatter):
+    ecs_version = '8.11.0'
+    __schema_dict = {
+        'ECS_VERSION': 'ecs.version',
+        'MESSAGE_TYPE': 'event.module',
+        'HOST': 'host.hostname',
+        'LOG_LEVEL': 'log.level',
+        'LOGGER_NAME': 'log.logger',
+        'LOG_SOURCE': 'log.syslog.hostname',
+        'LINE': 'log.origin.file.line',
+        'PATH': 'log.origin.file.name',
+        'FUNC_NAME': 'log.origin.function',
+        'STACK_TRACE': 'error.stack_trace',
+        'ERROR_TYPE': 'error.type',
+        'PROGRAM': 'process.executable',
+        'PROCESS_NAME': 'process.name',
+        'PID': 'process.pid',
+        'THREAD_NAME': 'process.thread.name',
+    }
+
+    formatter_logstash_message_field_list = (LogstashFormatter.formatter_logstash_message_field_list
```

Comment:
I wonder if it would be better to define this list explicitly in the constants module as the field names differ a lot from the legacy Logstash field list and so it might be worth to tell the user explicitly instead of just concatenating the legacy and ECS field name lists here.

Reply:
Makes sense.  
Maybe it would be even better to make all dotted fields top-level by default (with ability to override in constants)?

Reply:
I'm not sure I understand what you mean exactly especially with regard to the ability to override.

What I meant in the comment above was to add a new constant like `FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST` and specify the field list used by the ECS formatter as those fields are quite different from the legacy fields.
This would make it easier for users to modify the list of fields included in the event when using the ECS formatter.

Reply:
Isn't `*_FIELD_LIST` meant to define which fields should be top-level, not which fields to include?

Anyway, I agree that it's a good idea to define `*_FIELD_LIST` (not sure about `*_SKIP_LIST`, not many changes there) for ECS.
My idea of defining the default `FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST` would look like this:
```py
ALL_DOTTED_FIELDS = object()
FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST = ['@timestamp', '@version', ALL_DOTTED_FIELDS]
...
field_skip_set = self.formatter_logstash_message_field_set | {self._extra_prefix}
for key in list(message):
    if key in field_skip_set:
        continue
    if '.' in key and ALL_DOTTED_FIELDS in field_skip_set:
        continue
    message[self._extra_prefix][key] = message.pop(key)
```

Reply:
> Isn't `*_FIELD_LIST` meant to define which fields should be top-level, not which fields to include?

`FORMATTER_LOGSTASH_MESSAGE_FIELD_LIST` is the list of field names to be included in the Logstash event.
`FORMATTER_RECORD_FIELD_SKIP_LIST` is the list of field names of the *Python* log record which should *not* be included in the Logstash event. So this one doesn't need any changes for ECS support.

> Anyway, I agree that it's a good idea to define `*_FIELD_LIST` (not sure about `*_SKIP_LIST`, not many changes there) for ECS. My idea of defining the default `FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST` would look like this:
> 
> ```python
> ALL_DOTTED_FIELDS = object()
> FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST = ['@timestamp', '@version', ALL_DOTTED_FIELDS]
> ...
> field_skip_set = self.formatter_logstash_message_field_set | {self._extra_prefix}
> for key in list(message):
>     if key in field_skip_set:
>         continue
>     if '.' in key and ALL_DOTTED_FIELDS in field_skip_set:
>         continue
>     message[self._extra_prefix][key] = message.pop(key)
> ```

Why do you want to generate the list?
Can't we define it manually like `FORMATTER_LOGSTASH_MESSAGE_FIELD_LIST`, e.g.:
```
    FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST = [
        '@timestamp', '@version', 'host.hostname', 'log.level', 'log.syslog.hostname', ...]
```
?

Reply:
Looking at the implementation, `FORMATTER_LOGSTASH_MESSAGE_FIELD_LIST` seems to only be used to decide whether to put a field into `extra` or top-level, not to include/exclude it altogether:
https://github.com/eht16/python-logstash-async/blob/c60580b69103a945573586f3ee9bf09d16778f0b/logstash_async/formatter.py#L192-L205

<br>  

>Why do you want to generate the list?  

There's a lot of fields in ECS spec, most of them (all of them?) are dotted, so it looks to me like a sensible default. 
I don't see any immediate disadvantages to this, and it may even be slightly better for performance because list/set is smaller.
That said, I don't have a very strong opinion on this, we can do it manually. But it's probably going to be a long list, especially if we include fields that are not used by the lib, but may be used in user subclasses.

Reply:
Oops, you are completely right.
Sorry I should have checked the code before, in my memories the `FORMATTER_LOGSTASH_MESSAGE_FIELD_LIST` was meant to remove missing fields but it is actually not.

So, since `FORMATTER_LOGSTASH_MESSAGE_FIELD_LIST` defines only the location of the fields in the Logstash event, this makes not much sense for ECS as the structure is pretty much defined by ECS, making it less useful to redefine by the user.

Reply:
Probably so. 
We'd need to think how to differentiate extras then, since with current implementation `_get_extra_fields` could return both top-level ecs fields and true extra fields which should/could be under the prefix.

Reply:
So, I've added `FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST` which contains only ECS base fields (non dotted ones) and dotted fields are derived from formatter's schema.
I think we can keep `FORMATTER_LOGSTASH_ECS_MESSAGE_FIELD_LIST` in constants just in case a user wants to do something weird.